---
title: "Aedes aegypti Brazil - Local adaptation analysis - pcadapt"
author: "Luciano V Cosme"
date: "`r Sys.Date()`"
output:
  html_document:
    highlight: breezedark
    css:
      - "styles.css"
    toc: yes
    toc_float: no
    toc_depth: 5
editor_options:
  markdown:
    wrap: 120
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  eval                        = TRUE,
  echo                        = TRUE,
  cache                       = TRUE, # tidy = TRUE,
  class.output                = "bg-success"
)
knitr::opts_knit$set(
  root.dir = rprojroot::find_rstudio_root_file()
)
```


<span class="rainbow-title">Analysis code</span>

<!-- Custom JavaScript to apply the rainbow effect to the title -->
<script>
document.addEventListener("DOMContentLoaded", function() {
  var titleElements = document.querySelectorAll('h1');
  if (titleElements.length > 0) {
    titleElements[0].classList.add('rainbow-title');
  }
});
</script>


In this RMarkdown we will overview how to run local adaptation analysis with R.Sambada

## 1. R libraries and software

```{r libraries, message=FALSE, results='hide', warning=FALSE}
library(tidyverse)
library(colorout)
library(here)
library(ggplot2)
library(scales)
library(extrafont)
library(forcats)
library(ggrepel)
library(ggtext)
library(R.SamBada)
library(rgdal)
library(stats)
library(geosphere)
library(pcadapt)
```


## 2. pcadapt

We can the same data we used with outflank at output/local_adaptation/outflank/brazil



pcadapt performs principal component analysis and computes p-values to test for outliers. The test for outliers is based on the correlations between genetic variation and the first K principal components. pcadapt also handles Pool-seq data for which the statistical analysis is performed on the genetic markers frequencies. Returns an object of class pcadapt.


```{r}
# read the file
x <- read.pcadapt(here("output","local_adaptation", "outflank","brazil.bed"), type = "bed")
```


```{r, warning=FALSE}
# Run pcadapt with a large number of PCs
pcadapt_res <- pcadapt(x, K = 35)

# Create a scree plot
plot(pcadapt_res, option = "screeplot")
```

Choose PCs
```{r}
# perform the analysis
pcadapt_res <- pcadapt(
  x, 
  method = c("mahalanobis"),
  min.maf = 0.1,
  LD.clumping = NULL,
  tol = 1e-04,
  K = 2) # choose the right K for your data

str(pcadapt_res)
```



```{r}
#   ____________________________________________________________________________
#   import the bim file with the SNP data                                   ####
snps <-
  read_delim(                    # to learn about the options use here, run ?read_delim on the console.
    here(
      "output", "local_adaptation", "outflank", "brazil.bim"
    ),                           # use library here to load it
    col_names      = FALSE,      # we don't have header in the input file
    show_col_types = FALSE,      # suppress message from read_delim
    col_types      = "ccidcc"    # set the class of each column
  )
#
# set column names
colnames(
  snps
) <-                             # to add a header in our tibble
  c(
    "Scaffold", "SNP", "Cm", "Position", "Allele1", "Allele2"
  )
#
# check the tibble
head(snps)
```

```{r}
# Combine into a data frame
pcadapt_res_df <- data.frame(
  SNP = snps$SNP,
  Chromosome = snps$Scaffold,
  Position = snps$Position,
  pvalues = pcadapt_res$pvalues,
  stat = pcadapt_res$stat
) |> drop_na()
head(pcadapt_res_df)
```


Adjust
```{r}
# Adjust the p-values using Benjamini-Hochberg method
padj <- p.adjust(pcadapt_res_df$pvalues,method="BH")

# Define significance level
alpha <- 0.05

# Get indices of significant SNPs
outliers_indices <- which(padj < alpha)

# Get the SNP IDs of the outliers
outlier_snps <- pcadapt_res_df$SNP[outliers_indices]

# Print the SNP IDs
length(outlier_snps)

# Save it
write.table(
  outlier_snps,
  file = here("output", "local_adaptation", "pcadapt","brazil_SNPs_pcadapt.txt"),
  row.names = FALSE,
  quote = FALSE,
  col.names = FALSE,
  sep = "\n"
)
```


```{r}
# Set colors for the chromosomes
color_vector <- c("#CCF6D6", "#F6E1CC", "#CCD8F6")  # Add or remove colors as needed.
names(color_vector) <- unique(pcadapt_res_df$Chromosome)  # Make sure unique(df_sub$Chromosome) gives all unique Chromosome values.

# Adjust the p-values using Benjamini-Hochberg method
padj <- p.adjust(pcadapt_res_df$pvalues,method="BH")

# Define significance level
alpha <- 0.05

# Get indices of significant SNPs
outliers_indices <- which(padj < alpha)

# Create a new 'highlight' column, initially set to FALSE
pcadapt_res_df$highlight <- FALSE

# Update the 'highlight' column to highlight these outlier SNPs
pcadapt_res_df$highlight[outliers_indices] <- TRUE

# Identify the SNP with the smallest p-value in each chromosome among the associated SNPs
min_pval_snps <- pcadapt_res_df |>
  filter(highlight) |>
  group_by(Chromosome) |>
  slice(which.min(pvalues))

# Custom label function
k_format <- function(x) {
  paste0(format(x / 1e6, big.mark = "", scientific = FALSE), "k")
}

# source the plotting function
source(
  here(
    "scripts", "analysis", "my_theme2.R"
  )
)

# Create the plot
ggplot(pcadapt_res_df, aes(x = Position, y = -log10(pvalues))) +
  geom_point(aes(color = Chromosome),
             data = subset(pcadapt_res_df, highlight == FALSE),
             size = .5) +
  geom_point(
    color = "magenta",
    data = subset(pcadapt_res_df, highlight == TRUE),
    size = .5
  ) +
  scale_color_manual(values = color_vector, guide = "none") +
  labs(title = "pcadapt Brazil", x = "Position", y = "-log10(p-value)", caption = "SNPs highlighted in magenta are significant after Benjamini-Hochberg adjustment (p<0.05).") +
  facet_wrap(~ Chromosome, scales = "free_x") +
  scale_x_continuous(labels = k_format) +
  my_theme() +
  theme(panel.spacing = unit(.2, "lines"),
        plot.margin = unit(c(1, 1, 2, 2), "lines"),
        plot.caption = element_markdown(face = "italic", color = "#574E4E")) +
  # Annotate only the SNP with the smallest p-value for each chromosome among the associated SNPs
  geom_text_repel(
    data = min_pval_snps,
    aes(label = SNP),
    size = 3,
    nudge_y = 0.01,
    segment.color = NA,
    max.overlaps = Inf # Set to 'Inf' to show all annotations
  )

# Save the plot
ggsave(
  here(
    "output", "local_adaptation", "figures", "pcadapt_brazil.pdf"
  ),
  width  = 8,
  height = 5,
  units  = "in"
)
```

```{r}
plot(pcadapt_res, option = "qqplot")
```


```{r}
plot(pcadapt_res, option = "scores")
```

```{r}
hist(pcadapt_res$pvalues, xlab = "p-values", main = NULL, breaks = 50, col = "pink")
```


```{r, warning=FALSE}
plot(pcadapt_res, option = "stat.distribution")
```

Choosing a cutoff for outlier detection

```{r}
library(qvalue)
```

q-values
```{r}
qval <- qvalue(pcadapt_res$pvalues)$qvalues
alpha <- 0.05
outliers <- which(qval < alpha)
length(outliers)
```

Benjamini-Hochberg Procedure
```{r}
padj <- p.adjust(pcadapt_res$pvalues,method="BH")
alpha <- 0.05
outliers <- which(padj < alpha)
length(outliers)
```

Bonferroni correction

```{r}
padj <- p.adjust(pcadapt_res$pvalues,method="bonferroni")
alpha <- 0.05
outliers <- which(padj < alpha)
length(outliers)
```

Loadings

```{r}
par(mfrow = c(2, 2))
for (i in 1:2)
  plot(pcadapt_res$loadings[, i], pch = 19, cex = .3, ylab = paste0("Loadings PC", i))
```


## 3. Venn diagram outFlank vs pcadapt

```{r}
library(ggvenn)
# Read data from txt files
outflank_snps <-
  read.table(
    here("output", "local_adaptation", "outflank", "brazil_SNPs_outFlank.txt"),
    stringsAsFactors = FALSE
  ) |>
  drop_na()
pcadapt_snps <-
  read.table(
    here("output", "local_adaptation", "pcadapt", "brazil_SNPs_pcadapt.txt"),
    stringsAsFactors = FALSE
  )

# Create a list with all dataframes
list_of_clusters <- list("outFlank" = outflank_snps$V1, "pcadapt" = pcadapt_snps$V1)


# Create Venn diagram
venn_diagram <- ggvenn(list_of_clusters, fill_color = c("steelblue", "darkorange"))
print(venn_diagram)

# Find common SNPs
common_SNPs <- Reduce(intersect, list_of_clusters)

# Save the shared SNP ids into a txt file
write.table(
  common_SNPs,
  file = here("output", "local_adaptation", "common_SNPs_pcadapt_outflank.txt"),
  row.names = FALSE,
  col.names = FALSE,
  quote = FALSE
)

# # Save Venn diagram to PDF
output_path <- here("output", "local_adaptation", "figures", "significant_snps.pdf")
ggsave(output_path, venn_diagram, height = 5, width = 5, dpi = 300)
```

Create plot with the shared SNPs
```{r}
# Filter the dataframe
filtered_df <- pcadapt_res_df |> filter(SNP %in% common_SNPs)

# Create the plot
ggplot(pcadapt_res_df, aes(x = Position, y = -log10(pvalues))) +
  geom_point(
    aes(color = Chromosome),
    data = subset(pcadapt_res_df, highlight == FALSE),
    size = .3
  ) +
  geom_point(
    color = "magenta",
    data = subset(filtered_df, highlight == TRUE),
    size = .3
  ) +
  scale_color_manual(values = color_vector, guide = "none") +
  labs(x = "Position", y = "-log10(p-value)", caption = "SNPs highlighted in magenta are significant after \n Benjamini-Hochberg adjustment (p<0.001) with pcadapt, \nand were identified as outliers with outFlank") +
  facet_wrap( ~ Chromosome, scales = "free_x") +
  scale_x_continuous(labels = k_format) +
  my_theme() +
  theme(
    panel.spacing = unit(.2, "lines"),
    plot.margin = unit(c(1, 1, 2, 2), "lines"),
    plot.caption = element_text(face = "italic")
  ) +
  # Annotate only the SNP with the smallest p-value for each chromosome among the associated SNPs
  geom_text_repel(
    data = min_pval_snps,
    aes(label = SNP),
    size = 3,
    nudge_y = 0.01,
    segment.color = NA,
    max.overlaps = Inf # Set to 'Inf' to show all annotations
  )
```

```{r}
# Save the plot
ggsave(
  here(
    "output", "local_adaptation", "figures", "pcadapt_outflank.pdf"
  ),
  width  = 8,
  height = 5,
  units  = "in"
)
```

