---
title: "VGSC - Genotype frequencies."
author: "Luciano V Cosme"
date: "`r Sys.Date()`"
output:
  html_document:
    highlight: breezedark
    css:
      - "styles.css"
    toc: yes
    toc_float: no
    toc_depth: 5
editor_options:
  markdown:
    wrap: 120
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  eval                        = TRUE,
  echo                        = TRUE,
  cache                       = TRUE, # tidy = TRUE,
  class.output                = "bg-success"
)
knitr::opts_knit$set(
  root.dir = rprojroot::find_rstudio_root_file()
)
```

# <span class="rainbow-title">Analysis code</span>

<!-- Custom JavaScript to apply the rainbow effect to the title -->
<script>
document.addEventListener("DOMContentLoaded", function() {
  var titleElements = document.querySelectorAll('h1');
  if (titleElements.length > 0) {
    titleElements[0].classList.add('rainbow-title');
  }
});
</script>


## 1. R libraries and software

```{r libraries, message=FALSE, results='hide'}
library(tidyverse)
library(here)
library(dplyr)
library(ggplot2)
library(colorout)
library(extrafont)
library(scales)
library(stringr)
library(flextable)
library(ggtext)
library(RColorBrewer)
library(ggrepel)
library(grid)
library(forcats)
library(officer)
library(viridis)
library(gridExtra)
```


## 2. Frequency of SNPs within the sodium channel (VGSC)
The gene
AAEL023266 chromosome 3 315926360 - 316405639 

We can find out if we have SNPs within this genomic regions


```{r}
# load the function that we saved earlier
source(
  here(
    "scripts", "analysis", "import_bim.R"
  ),
  local = knitr::knit_global()
)

# import the file
file10 <- import_bim(
  here(
    "output", "global_brazil", "file10.bim"
  )
) |>
  dplyr::rename(Chromosome = Scaffold)

head(file10)
```

Now we can check if there are SNPs between the VGSC gene

```{r}
# Filter for Chromosome 3 and positions between 315926360 - 316405639
filtered_data <- file10 %>%
  filter(Chromosome == "3", Position >= 315926360, Position <= 316405639)

# View the filtered data
print(filtered_data)
```

We have two SNPs and we can estimate their genotype frequency now

```{r}
# Save it
write.table(
  filtered_data$SNP,
  file = here("output", "frequencies", "vgsc_snps.txt"),
  row.names = FALSE,
  quote = FALSE,
  col.names = FALSE,
  sep = "\n"
)

```


```{bash}
for fam in $(awk '{print $1}' output/global_brazil/file10.fam | sort | uniq); 
do 
echo $fam | \
plink2 \
--extract output/frequencies/vgsc_snps.txt \
--bfile output/global_brazil/file10 \
--keep-fam /dev/stdin \
--out output/frequencies/vgsc/$fam \
--geno-counts \
--silent
done
```


Before run the code below, remember to delete the ID.txt in the directory otherwise it will give you a error

Get genotype frequencies
```{bash}
# Create a directory for the SNP genotype files if it doesn't exist
mkdir -p output/frequencies/vgsc/genotype_files

# Create a list of unique SNPs
unique_snps=$(awk 'NR>1 {print $2}' output/frequencies/vgsc/*.gcount | sort | uniq)

# For each unique SNP
for snp in $unique_snps; do
    # Create an empty temporary file
    > output/frequencies/vgsc/genotype_files/${snp}_tmp.txt
  
    # Loop through each family's genotype count file
    for file in output/frequencies/vgsc/*.gcount; do
        family=$(basename $file .gcount)  # Extract the family name from the filename
        
        # Extract alleles for the SNP from the current family file
        ref_allele=$(awk -v snp="$snp" 'NR>1 && $2 == snp {print $3}' $file | head -n 1)
        alt_allele=$(awk -v snp="$snp" 'NR>1 && $2 == snp {print $4}' $file | head -n 1)
        genotypes="${ref_allele}${ref_allele}\t${ref_allele}${alt_allele}\t${alt_allele}${alt_allele}"

        # Extract genotype frequencies for the SNP from the current family file and append to the SNP's file
        awk -v snp="$snp" -v family="$family" 'NR>1 && $2 == snp {
            total_count = $5 + $6 + $7
            ref_ref_freq = $5 / total_count
            ref_alt_freq = $6 / total_count
            alt_alt_freq = $7 / total_count
            print $2 "\t" family "\t" ref_ref_freq "\t" ref_alt_freq "\t" alt_alt_freq
        }' $file >> output/frequencies/vgsc/genotype_files/${snp}_tmp.txt

        # If the SNP genotype file does not exist yet, create it with a header using the appropriate alleles
        if [[ ! -f output/frequencies/vgsc/genotype_files/${snp}.txt ]]; then
            echo -e "SNP\tStratum\t$genotypes" > output/frequencies/vgsc/genotype_files/${snp}.txt
        fi
    done

    # Append the extracted data to the SNP genotype file
    cat output/frequencies/vgsc/genotype_files/${snp}_tmp.txt >> output/frequencies/vgsc/genotype_files/${snp}.txt
    rm output/frequencies/vgsc/genotype_files/${snp}_tmp.txt
done
```

Delete the id file
```{bash}
rm output/frequencies/vgsc/genotype_files/ID.txt
```


Now we can prepare the data for plotting
```{r}
files <- list.files(path = here("output", "frequencies", "vgsc", "genotype_files"), pattern = "*.txt", full.names = TRUE)

all_data <- lapply(files, function(file) {
  dat <- read.table(
    file,
    header = TRUE,
    sep = "\t",
    stringsAsFactors = FALSE,
    colClasses = c("character")
  )
  
  # Storing genotype names from columns 3 to 5
  genotype_names <- colnames(dat)[3:5]
  
  # Renaming columns 3 to 5
  colnames(dat)[3:5] <- c("Genotype1_Value", "Genotype2_Value", "Genotype3_Value")
  
  # Add the genotype names as new columns
  dat$Genotype1_Name <- genotype_names[1]
  dat$Genotype2_Name <- genotype_names[2]
  dat$Genotype3_Name <- genotype_names[3]
  
  # Add SNP column
  dat$SNP <- gsub(".txt", "", basename(file))
  
  return(dat)
})

# Binding all data frames
all_data_df <- do.call(rbind, all_data)


# # Update names
# all_data_df <- all_data_df %>%
#   mutate(Stratum = case_when(
#     Stratum == "Background_Female_Dead" ~ "BFD",
#     Stratum == "Background_Male_Dead"   ~ "BMD",
#     Stratum == "Selection_Female_Alive" ~ "SFA",
#     Stratum == "Selection_Male_Alive"   ~ "SMA",
#     TRUE                                ~ Stratum  # Keep original value for other cases
#   ))
```



Find all genotypes
```{r}
# Combine all genotype name columns into one vector and find unique values
unique_genotypes <- unique(c(all_data_df$Genotype1_Name, all_data_df$Genotype2_Name, all_data_df$Genotype3_Name))

# Print the unique genotypes
print(unique_genotypes)
```

Set the colors using Rcolorbrew library
```{r}
# Generate a palette with viridis
genotype_palette <- viridis::viridis(10)

# First get the maximum from one palette
palette1 <- brewer.pal(11, "Spectral")
# Then get the rest from another palette, making sure to not have duplicates
palette2 <- brewer.pal(5, "Set3")

# Combine them while excluding any duplicates
genotype_colors <- unique(c(palette1, palette2))

# colors
genotype_colors <- genotype_colors[1:6]
```


Plot it
```{r, fig.height=5, fig.width=15}
# colors
genotype_colors <- c("red", "gray", "cyan", "orange", "magenta", "green")

# Reshape the data to long format for plotting genotypes
geno_long <- all_data_df %>%
  mutate(
    Genotype1_Name = as.character(Genotype1_Name),
    Genotype2_Name = as.character(Genotype2_Name),
    Genotype3_Name = as.character(Genotype3_Name)
  ) %>%
  pivot_longer(
    cols = c(Genotype1_Value, Genotype2_Value, Genotype3_Value),
    names_to = "Genotype_Num",
    values_to = "Value"
  ) %>%
  mutate(
    Genotype = case_when(
      Genotype_Num == "Genotype1_Value" ~ Genotype1_Name,
      Genotype_Num == "Genotype2_Value" ~ Genotype2_Name,
      Genotype_Num == "Genotype3_Value" ~ Genotype3_Name
    ),
    Value = as.numeric(Value)  # Ensure that Value is numeric
  )

# # Define the colors for the borders corresponding to each Stratum
# stratum_border_colors <- c("SFA" = "red", "SMA"= "red", "BFD" = "black", "BMD" = "black")

# load plotting theme
source(
  here(
    "scripts", "analysis", "my_theme2.R"
  )
)

# Update the plot with genotypes
plot <- ggplot(geno_long, aes(x = Stratum, y = Value, fill = Genotype, group = Genotype)) +
  geom_bar(stat = "identity", position = position_dodge(width = 0.9), aes(color = "gray"), linewidth = 0.1) +
  scale_fill_manual(name = "Genotype", values = genotype_colors) +
  scale_y_continuous(limits = c(0, 1.2), breaks = seq(0, 1, by = 0.25)) +
  facet_wrap(~ SNP, scales = "free_x", ncol = 1) +
  geom_text(
    aes(label = Genotype, y = Value + 0.02, group = Genotype),
    position = position_dodge(width = 0.9),
    vjust = -0.25,
    size = 1.8,
    check_overlap = TRUE
  ) +
  labs(x = "Population", y = "Frequency") +
  my_theme() +
  theme(
    strip.text.x = element_text(size = 10, face = "bold", margin = margin(t = 1, b = 1, unit = "pt")),
    legend.position = "top",
    legend.title.align = 0.5, 
    strip.background = element_rect(fill = "#e8e8e8", colour = NA),
    panel.spacing = unit(1, "lines"),
    axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1, color = "black"),
    axis.text.y = element_text(color = "black")
  ) +
  guides(
    fill = guide_legend(nrow = 1, byrow = TRUE, title = "Genotypes"),
    color = "none"
  ) 

# Add ggtext compatible theme settings for using HTML in axis.text.x
plot <- plot + theme(axis.text.x = element_markdown())

# Output the plot
print(plot)
```

Import the mata data
```{r}
cities <- readRDS(
  here(
    "output", "global_brazil", "cities_loc.rds"
  )
)
head(cities)
```

Merge the objects
```{r, fig.height=5, fig.width=13}
# Merging the datasets
merged_data <- merge(geno_long, cities, by.x = "Stratum", by.y = "pop")

# Arrange the merged data by Continent and then by Country
arranged_data <- merged_data %>%
  arrange(Continent, country)

# Create an auxiliary ordering column
arranged_data$Order <- as.numeric(factor(paste(arranged_data$Continent, arranged_data$country)))

# Create the plot with the new ordering
plot <- ggplot(arranged_data, aes(x = reorder(Stratum, Order), y = Value, fill = Genotype)) +
  geom_bar(stat = "identity", position = position_dodge()) +
  scale_fill_manual(name = "Genotype", values = genotype_colors) +
  scale_y_continuous(limits = c(0, 1.2), breaks = seq(0, 1, by = 0.25)) +
  facet_wrap(~ SNP, scales = "free_x", ncol = 1) +
  geom_text(
    aes(label = Genotype, y = Value + 0.02, group = Genotype),
    position = position_dodge(width = 0.9),
    vjust = -0.25,
    size = 1.8,
    check_overlap = TRUE
  ) +
  labs(x = "Population", y = "Frequency") +
  my_theme() +
  theme(
    strip.text.x = element_text(size = 10, face = "bold", margin = margin(t = 1, b = 1, unit = "pt")),
    legend.position = "top",
    legend.title.align = 0.5, 
    strip.background = element_rect(fill = "#e8e8e8", colour = NA),
    panel.spacing = unit(1, "lines"),
    axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1, color = "black"),
    axis.text.y = element_text(color = "black")
  ) +
  guides(
    fill = guide_legend(nrow = 1, byrow = TRUE, title = "Genotypes"),
    color = "none"
  )

# Print the plot
print(plot)

# Save it
output_path <- here("output", "frequencies", "figures", "vgsc_SNPs_global.pdf")
ggsave(output_path, plot, height = 5, width = 15, dpi = 300)
```
With facets by continent
```{r, fig.height=8, fig.width=10}
# Subset the data for the first SNP
snp1_data <- arranged_data[arranged_data$SNP == "AX-93235430", ]

# Ensure that there are 16 unique colors
genotype_colors <- c("red", "black", "blue")

# Create the plot with faceting by continent
plot <- ggplot(snp1_data, aes(x = reorder(Stratum, Order), y = Value, fill = Genotype)) +
  geom_bar(stat = "identity", position = position_dodge()) +
  scale_fill_manual(name = "Genotype", values = genotype_colors) +
  scale_y_continuous(limits = c(0, 1.2), breaks = seq(0, 1, by = 0.25)) +
  facet_wrap(~ Continent, scales = "free_x", ncol = 1) +  # Faceting by continent
  geom_text(
    aes(label = Genotype, y = Value + 0.02, group = Genotype),
    position = position_dodge(width = 0.9),
    vjust = -0.25,
    size = 1.8,
    check_overlap = TRUE
  ) +
  labs(x = "Population", y = "Frequency") +
  my_theme() +
  theme(
    strip.text.x = element_text(size = 10, face = "bold", margin = margin(t = 1, b = 1, unit = "pt")),
    legend.position = "top",
    legend.title.align = 0.5, 
    strip.background = element_rect(fill = "#e8e8e8", colour = NA),
    panel.spacing = unit(1, "lines"),
    axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1, color = "black"),
    axis.text.y = element_text(color = "black")
  ) +
  guides(
    fill = guide_legend(nrow = 1, byrow = TRUE, title = "Genotypes"),
    color = "none"
  )

# Print the plot
print(plot)
```


With facets by continent
```{r, fig.height=8, fig.width=14}
# Subset the data for the first SNP
snp1_data <- arranged_data[arranged_data$SNP == "AX-93243399", ]

# Ensure that there are 16 unique colors
genotype_colors <- c("red", "black", "blue")

# Create the plot with faceting by continent
plot <- ggplot(snp1_data, aes(x = reorder(Stratum, Order), y = Value, fill = Genotype)) +
  geom_bar(stat = "identity", position = position_dodge()) +
  scale_fill_manual(name = "Genotype", values = genotype_colors) +
  scale_y_continuous(limits = c(0, 1.2), breaks = seq(0, 1, by = 0.25)) +
  facet_wrap(~ Continent, ncol = 1) +  # Faceting by continent
  geom_text(
    aes(label = Genotype, y = Value + 0.02, group = Genotype),
    position = position_dodge(width = 0.9),
    vjust = -0.25,
    size = 1.8,
    check_overlap = TRUE
  ) +
  labs(x = "Population", y = "Frequency") +
  my_theme() +
  theme(
    strip.text.x = element_text(size = 10, face = "bold", margin = margin(t = 1, b = 1, unit = "pt")),
    legend.position = "top",
    legend.title.align = 0.5, 
    strip.background = element_rect(fill = "#e8e8e8", colour = NA),
    panel.spacing = unit(1, "lines"),
    axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1, color = "black"),
    axis.text.y = element_text(color = "black")
  ) +
  guides(
    fill = guide_legend(nrow = 1, byrow = TRUE, title = "Genotypes"),
    color = "none"
  )

# Print the plot
print(plot)
```











