---
title: "Genomic adaptations of Aedes aegypti in Brazil"
author: "Luciano V Cosme"
date: "`r Sys.Date()`"
output:
  html_document:
    highlight: breezedark
    css:
      - "styles.css"
    toc: yes
    toc_float: no
    toc_depth: 5
editor_options:
  markdown:
    wrap: 120
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
  eval                        = TRUE,
  echo                        = TRUE,
  cache                       = TRUE, # tidy = TRUE,
  class.output                = "bg-success"
)
knitr::opts_knit$set(
  root.dir = rprojroot::find_rstudio_root_file()
)
```


# <span class="rainbow-title"></span>

<!-- Custom JavaScript to apply the rainbow effect to the title -->
<script>
document.addEventListener("DOMContentLoaded", function() {
  var titleElements = document.querySelectorAll('h1');
  if (titleElements.length > 0) {
    titleElements[0].classList.add('rainbow-title');
  }
});
</script>


## Author

Luciano V. Cosme

ðŸ“§ Email: [luciano.cosme@yale.edu](mailto:luciano.cosme@yale.edu)

## Overview

This repository accompanies our manuscript on the genetic diversity and adaptations of Aedes aegypti mosquitoes in Brazil. The repository includes the complete R project, code in RMD and HTML format. Our study aims to clarify the global population structure and genetic differentiation of Ae. aegypti, focusing on the unique ecological environments in Brazil. By genotyping 50k SNPs across 1,330 samples, we conducted comprehensive ancestry analyses, population structure evaluations, and genetic differentiation assessments. The selection scans and gene-environment association analyses revealed significant genetic diversity and local adaptation, identifying eight genetic clusters in Brazil. These clusters likely result from environmental pressures and reintroduction post-eradication efforts. Our findings emphasize the importance of considering local adaptation in vector control strategies.

## Contents

- [File S1: How we mapped probes to the AaegL5 genome assembly](01.mapping_probes.html)
- [File S2: Segregation analysis with laboratory crosses](02.Segregation_aegypti.html)
- [File S3: Quality control for the Brazilian data set](03.Quality_control_brazil.html)
- [File S4: Quality control for the global data set](04.Quality_Control_global_and_brazil_dataset.html)
- [File S5: Functional annotation of SNPs for downstream analyses](05.SNPs_functional_annotation.html)
- [File S6: Population structure analysis with fastStructure](06.fastStructure_analysis.html)
- [File S7: Population structure analysis with Admixture](07.Admixture_analysis.html)
- [File S8: Population structure analysis with LEA](08.LEA_analysis.html)
- [File S9: Interpolation of Q matrices with Tess3R](09.Interpolate_q_matrix_map.html)
- [File S10: Linkage analysis with PopLDdecay](10.Linkage.html)
- [File S11: Selection scan with OutFLANK](11.local_adaptation_outflank.html)
- [File S12: Selection scan with pcadapt](12.local_adaptation_pcadapt.html)
- [File S13:Local adaptation analysis with sambada and R.sambada](13.local_adaptation_sambada.html)
- [File S14: Fst analsyis](14.Fst_chip.html)
- [File S15: Migration analysis with BA3-SNPs](15.Migration_ba3_snps.html)
- [File S16: Genotype frequencies of SNPs on the VGSC](16.Frequencies.html)

## R project with all the data is availabe in Zenodo
[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.11390918.svg)](https://doi.org/10.5281/zenodo.11390918)


The study was supported by NIH grant UO1 AI115595.








